bpf/lib/nat.h

```diff
@@ -1157,6 +1308,83 @@ snat_v6_rev_nat_handle_mapping(struct __ctx_buff *ctx,
 		       NAT_PUNT_TO_STACK : DROP_NAT_NO_MAPPING;
 }
 
+static __always_inline int snat_v6_icmp_rewrite_embedded(struct __ctx_buff *ctx,
+							 struct ipv6_ct_tuple *tuple,
+							 struct ipv6_nat_entry *state,
+							 __u32 l4_off)
+{
+	struct csum_offset csum = {};
+	int hdrlen;
+	__u32 inner_l4_off;
+
+	hdrlen = ipv6_hdrlen(ctx, &tuple->nexthdr);
+	if (hdrlen < 0)
+		return hdrlen;
+
+	/* This inner_l4_off is L4 offset of inner packet, where l4_off is L4
+	 * offset of outer packet.
+	 *
+	 * NOTE: -4 because struct icmp6hdr includes echo
+	 * identifier which is not used for ICMP Error.
+	 *
+	 * l4_off                  inner_l4_off
+	 *  [ ICMP ERROR | INNER IP | INNER L4 ...]
+	 */
+	inner_l4_off = l4_off + sizeof(struct icmp6hdr) - 4 + hdrlen;
+
+	if (ipv6_addrcmp(&state->to_daddr, &tuple->daddr) == 0 &&
+	    state->to_dport == tuple->dport)
+		return 0;
+
+	csum_l4_offset_and_flags(tuple->nexthdr, &csum);
+	if (state->to_dport != tuple->dport) {
+		__be32 suml4 = 0;
+
+		switch (tuple->nexthdr) {
+		case IPPROTO_TCP:
+		case IPPROTO_UDP:
+			/* In case that the destination port has been NATed from
+			 * target to dest. We want the embedded packet which
+			 * should refer to endpoint dest going back to original.
+			 */
+			if (ctx_store_bytes(ctx, inner_l4_off + offsetof(struct tcphdr, source),
+					    &state->to_dport, sizeof(state->to_dport), 0) < 0)
+				return DROP_WRITE_ERROR;
+			break;
+		case IPPROTO_ICMPV6: {
+			/* In case that the ID has been used as source port during
+			 * NAT from target to dest. We want the embedded packet
+			 * which should refer to endpoint -> dest going back to
+			 * original.
+			 */
+			if (ctx_store_bytes(ctx, inner_l4_off +
+					    offsetof(struct icmp6hdr,
+						     icmp6_dataun.u_echo.identifier),
+					    &state->to_dport,
+					    sizeof(state->to_dport), 0) < 0)
+				return DROP_WRITE_ERROR;
+			break;
+		}
+		default:
+			return DROP_INVALID;
+		}
+		/* By recomputing L4 checksum of inner packet we avoid having
+		 * to recompute L4 of the ICMP Error.
+		 */
+		suml4 = csum_diff(&tuple->dport, 4, &state->to_dport, 4, 0);
+		if (csum_l4_replace(ctx, inner_l4_off, &csum, 0, suml4, BPF_F_PSEUDO_HDR) < 0)
+			return DROP_CSUM_L4;
+	}
+	/* Change IP of source address of inner packet to refer the
+	 * endpoint.
+	 */
+	if (ctx_store_bytes(ctx, l4_off + sizeof(struct icmp6hdr) - 4 +
+			    offsetof(struct ipv6hdr, saddr),
+			    &state->to_daddr, 16, 0) < 0)
+		return DROP_WRITE_ERROR;
```

Comment:
This is `inner_l4_off` and `ipv6_store_saddr()` - right?

Reply:
You mean I should use `inner_l4_off` and remove bytes to point on the IP headers?

Reply:
All good - I misread, it would be `inner_l3_off` here of course. Thanks!